### 1. Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

### Import libraries

In [87]:
import pandas as pd
import numpy as np
import os

### Import data

In [89]:
# Create path
path = r'C:\Users\nanh9\Instacart Basket Analysis'

In [90]:
# Importing orders_products_merged_derived.pkl File
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data' , 'Prepared Data' , 'orders_products_merged_busiest.pkl'))

In [91]:
# Check df.shape
ords_prods_merge.shape

(32399732, 19)

### 2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [93]:
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.667853
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

### 3. Analyse the result. How do the results for the entire dataframe differ from those of the subset?

- When comparing the subset to the entire dataset, the results now include all 'department_id' values across the complete dataframe.
- Moreover, the average values for the department IDs that were part of the subset have shifted.
- Some departments show an increase in their mean values, while others have decreased.
- For instance, in the subset, department ID 4 had a mean value of 18.825780, but within the entire dataset, this value decreased to 17.811403.

### 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [97]:
# Creating new column "max_order"
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\nanh9\AppData\Local\Temp\ipykernel_348\2324063168.py:2: FutureWarning: The provided callable <function max at 0x0000020E567363E0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [98]:
ords_prods_merge.head(20)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998  3139998            28                   6                 11   
1    1977647  1977647            30                   6                 17   
2     389851   389851             2                   0                 21   
3     652770   652770             1                   3                 13   
4    1813452  1813452             3                   4                 17   
5    1701441  1701441            16                   1                  7   
6    1871483  1871483             3                   2                 14   
7    1290456  1290456            12                   3                 10   
8     369558   369558            10                   1                 20   
9     589712   589712            15                   1                 12   
10   1798115  1798115             1                   5                 11   
11    497193   497193            36                   3                 13   
12   1727826  1727826            38                   4                 13   
13   2694698  2694698            39                   2                 12   
14   3171753  3171753             6                   3                 10   
15    186761   186761             7                   2                  8   
16    317121   317121             9                   5                  8   
17    200390   200390             4                   2                  0   
18    374436   374436             5                   1                 21   
19   2920553  2920553             6                   1                  0   

    days_since_last_order  first_order  add_to_cart_order  reordered _merge  \
0                     3.0        False                  5          0   both   
1                    20.0        False                  1          1   both   
2                     6.0        False                 20          0   both   
3                     NaN         True                 10          0   both   
4                     9.0        False                 11          1   both   
5                    26.0        False                  7          0   both   
6                    30.0        False                  2          0   both   
7                    30.0        False                  1          0   bo

In [99]:
# Assign loyalty flag
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [100]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'


In [101]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [102]:
# Check loyalty flags
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
New customer        15782942
Regular customer    13210772
Loyal customer       3406018
Name: count, dtype: int64

In [103]:
# Check Loyalty flag in dataframe
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number','max_order']].head(20)

user_id      loyalty_flag  order_number  max_order
0   3139998  Regular customer            28         28
1   1977647  Regular customer            30         30
2    389851      New customer             2          2
3    652770      New customer             1          1
4   1813452      New customer             3          3
5   1701441  Regular customer            16         16
6   1871483      New customer             3          3
7   1290456  Regular customer            12         12
8    369558      New customer            10         10
9    589712  Regular customer            15         15
10  1798115      New customer             1          1
11   497193  Regular customer            36         36
12  1727826  Regular customer            38         38
13  2694698  Regular customer            39         39
14  3171753      New customer             6          6
15   186761      New customer             7          7
16   317121      New customer             9          9
17   200390      New customer             4          4
18   374436      New customer             5          5
19  2920553      New customer             6          6

### 5.The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.


In [105]:
# the min max and mean of each loyalty flag
ords_prods_merge.groupby(['loyalty_flag']).agg({'prices':['mean','min','max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.775537  1.0  25.0
New customer      7.794679  1.0  25.0
Regular customer  7.790576  1.0  25.0

- New customer have the lowest price of products (avg & max)
- Loyal customer have the highest price of products (avg & max)

### 6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
- If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
- If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [108]:
# Creating column for average spender
ords_prods_merge['average_spend']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\nanh9\AppData\Local\Temp\ipykernel_348\655122787.py:2: FutureWarning: The provided callable <function mean at 0x0000020E56736DE0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['average_spend']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [109]:
# Creating flag for low and high spender
ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spending_flag']='low_spender'

ords_prods_merge.loc[ords_prods_merge['average_spend']>= 10, 'spending_flag'] = 'high_spender'

In [113]:
# Check spender flags
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
low_spender     30771345
high_spender     1628387
Name: count, dtype: int64

In [115]:
pd.options.display.max_rows = 100

In [116]:
ords_prods_merge[['user_id', 'average_spend', 'spending_flag', 'prices']].head(100)

user_id  average_spend spending_flag  prices
0   3139998       6.980000   low_spender     5.8
1   1977647       8.475000   low_spender     5.8
2    389851       7.909677   low_spender     5.8
3    652770       5.194737   low_spender     5.8
4   1813452       4.965000   low_spender     5.8
5   1701441       7.087500   low_spender     5.8
6   1871483       5.833333   low_spender     5.8
7   1290456       5.600000   low_spender     5.8
8    369558       6.700000   low_spender     5.8
9    589712       4.200000   low_spender     5.8
10  1798115       3.866667   low_spender     5.8
11   497193       5.266667   low_spender     5.8
12  1727826       5.500000   low_spender     5.8
13  2694698       5.300000   low_spender     5.8
14  3171753       5.433333   low_spender     5.8
15   186761       5.985714   low_spender     5.8
16   317121       6.340909   low_spender     5.8
17   200390       6.591667   low_spender     5.8
18   374436       5.610000   low_spender     5.8
19  2920553       5.610000   low_spender     5.8
20  2813049       6.792308   low_spender     5.8
21  2365998       6.187500   low_spender     5.8
22  1604797       7.600000   low_spender     5.8
23  2799665       5.790000   low_spender     5.8
24  2228851       6.394444   low_spender     5.8
25  2337422       5.955556   low_spender     5.8
26  2699008       6.266667   low_spender     5.8
27  1914830       6.838889   low_spender     5.8
28   783503       7.323529   low_spender     5.8
29  2500516       7.961111   low_spender     5.8
30   716845       6.600000   low_spender     5.8
31  1955535       6.468421   low_spender     5.8
32  1424793       5.623529   low_spender     5.8
33   749498       6.863636   low_spender     5.8
34  2434812       5.900000   low_spender     5.8
35  1531889       5.900000   low_spender     5.8
36   726947       7.837500   low_spender     5.8
37   627615       6.800000   low_spender     5.8
38  2507439       4.620000   low_spender     5.8
39  1152609       4.714286   low_spender     5.8
40  1048339       4.853846   low_spender     5.8
41  2818381       4.433333   low_spender     5.8
42  2722928       6.833333   low_spender     5.8
43  1482338       5.118182   low_spender     5.8
44  2311344       9.512500   low_spender     5.8
45   812943       5.791667   low_spender     5.8
46   256839       9.144444   low_spender     5.8
47  2192488       7.414286   low_spender     5.8
48   207661       8.200000   low_spender     5.8
49  1394659       6.309091   low_spender     5.8
50  2353526       6.369231   low_spender     5.8
51  2269226       6.185000   low_spender     5.8
52   367316       6.393333   low_spender     5.8
53  1448367       6.054545   low_spender     5.8
54    19989       7.250000   low_spender     5.8
55  2735922       5.200000   low_spender     5.8
56  2469422       7.078571   low_spender     5.8
57  3185558       5.860000   low_spender     5.8
58   350540       6.511111   low_spender     5.8
59   513100       4.914286   low_spender     5.8
60  2081405       4.550000   low_spender     5.8
61  2209807       8.250000   low_spender     5.8
62   921798       8.233333   low_spender     5.8
63  2540226       9.580000   low_spender     5.8
64  2265792       4.900000   low_spender     5.8
65  3357714       5.800000   low_spender     5.8
66  1186573       5.800000   low_spender     5.8
67  3209740       4.350000   low_spender     5.8
68   138748       5.133333   low_spender     5.8
69  2027157       7.731250   low_spender     5.8
70  3052954       4.233333   low_spender     5.8
71  1486756       7.333333   low_spender     5.8
72   632514       9.900000   low_spender     5.8
73  1253470       5.800000   low_spender     5.8
74   837303       7.886667   low_spender     5.8
75  2734881       5.280000   low_spender     5.8
76  2164394       5.923077   low_spender     5.8
77  3354551       6.192857   low_spender     5.8
78  1381165       5.633333   low_spender     5.8
79  2809438       6.016129   low_spender     5.8
80   425352       7.060000   low_sp

### 7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
- If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
- If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
- If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [118]:
# Create a column for frequet users
ords_prods_merge['Customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

C:\Users\nanh9\AppData\Local\Temp\ipykernel_348\592977977.py:2: FutureWarning: The provided callable <function median at 0x0000020E56865B20> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['Customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)


In [119]:
# Creating flag on order frequency
ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] > 20, 'frequency_flag'] = 'Non_frequent_customer'

ords_prods_merge.loc[(ords_prods_merge['Customer_frequency']>10) & (ords_prods_merge['Customer_frequency']<=20),'frequency_flag'] = 'Regular customer'

ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent Customer'

In [120]:
ords_prods_merge.shape

(32399732, 25)

In [121]:
# Check frequency flags

ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent Customer        19039687
Regular customer          6054742
Non_frequent_customer     5229558
NaN                       2075745
Name: count, dtype: int64

NaN values tells us that there are 5 customers who have only bought once.

In [123]:
ords_prods_merge[['user_id','days_since_last_order', 'Customer_frequency', 'frequency_flag']].head(100)

user_id  days_since_last_order  Customer_frequency         frequency_flag
0   3139998                    3.0                 3.0      Frequent Customer
1   1977647                   20.0                20.0       Regular customer
2    389851                    6.0                 6.0      Frequent Customer
3    652770                    NaN                 NaN                    NaN
4   1813452                    9.0                 9.0      Frequent Customer
5   1701441                   26.0                26.0  Non_frequent_customer
6   1871483                   30.0                30.0  Non_frequent_customer
7   1290456                   30.0                30.0  Non_frequent_customer
8    369558                   19.0                19.0       Regular customer
9    589712                   15.0                15.0       Regular customer
10  1798115                    NaN                 NaN                    NaN
11   497193                    7.0                 7.0      Frequent Customer
12  1727826                    7.0                 7.0      Frequent Customer
13  2694698                   12.0                12.0       Regular customer
14  3171753                    2.0                 2.0      Frequent Customer
15   186761                    6.0                 6.0      Frequent Customer
16   317121                   10.0                10.0      Frequent Customer
17   200390                    7.0                 7.0      Frequent Customer
18   374436                    6.0                 6.0      Frequent Customer
19  2920553                   14.0                14.0       Regular customer
20  2813049                    8.0                 8.0      Frequent Customer
21  2365998                    6.0                 6.0      Frequent Customer
22  1604797                   28.0                28.0  Non_frequent_customer
23  2799665                    7.0                 7.0      Frequent Customer
24  2228851                    7.0                 7.0      Frequent Customer
25  2337422                    0.0                 0.0      Frequent Customer
26  2699008                    7.0                 7.0      Frequent Customer
27  1914830                    7.0                 7.0      Frequent Customer
28   783503                    7.0                 7.0      Frequent Customer
29  2500516                    7.0                 7.0      Frequent Customer
30   716845                    3.0                 3.0      Frequent Customer
31  1955535                   16.0                16.0       Regular customer
32  1424793                    5.0                 5.0      Frequent Customer
33   749498                    6.0                 6.0      Frequent Customer
34  2434812                    NaN                 NaN                    NaN
35  1531889                   21.0                21.0  Non_frequent_customer
36   726947                    2.0                 2.0      Frequent Customer
37   627615                    NaN                 NaN                    NaN
38  2507439                   18.0                18.0       Regular customer
39  1152609                    8.0                 8.0      Frequent Customer
40  1048339                   30.0                30.0  Non_frequent_customer
41  2818381                   25.0                25.0  Non_frequent_customer
42  2722928                    NaN                 NaN                    NaN
43  1482338                    9.0                 9.0      Frequent Customer
44  2311344                   10.0                10.0      Frequent Customer
45   812943                    7.0                 7.0      Frequent Customer
46   256839                    NaN                 NaN                    NaN
47  2192488                   30.0                30.0  Non_frequent_customer
48   207661                    NaN                 NaN                    NaN
49  1394659                    7.0                 7.0      Frequent Customer
50  2353526               

### Export data

In [125]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'aggregated_ords_prods_merge.pkl'))